Improve the block transforms to include also anchors when resolving uid back and forth

sneridagh opened the pull request:
Now it covers the use case if an anchor is present.

mister-roboto commented:
@sneridagh thanks for creating this Pull Request and helping to improve Plone!

TL;DR: Finish pushing changes, pass all other checks, then paste a comment:
```
@jenkins-plone-org please run jobs
```

To ensure that these changes do not break other parts of Plone, the Plone test suite matrix needs to pass, but it takes 30-60 min.  Other CI checks are usually much faster and the Plone Jenkins resources are limited, so when done pushing changes and all other checks pass either [start all Jenkins PR jobs yourself](https://jenkinsploneorg.readthedocs.io/en/latest/run-pull-request-jobs.html#run-a-pull-request-job), or simply add the comment above in this PR to start all the jobs automatically.

Happy hacking!

sneridagh commented:
@jenkins-plone-org please run jobs

davisagli commented on src/plone/restapi/deserializer/utils.py:
```diff
@@ -25,11 +25,13 @@ def path2uid(context, link):
+    suffix = ""
+    for suffix_separator in SUFFIXES:
+        if suffix_separator in path:
+            path, suffix = path.split(suffix_separator, 1)
```
@sneridagh You are overwriting `suffix` here. So if both separators are present, only one of the suffixes will be kept.

sneridagh replied:
It seemed wrong, but it's fine, I added a test. Ported the pattern to the serializer too.

davisagli commented on src/plone/restapi/serializer/utils.py:
```diff
@@ -23,13 +23,15 @@ def resolve_uid(path):
         return path, None
     href = brain.getURL()
     if suffix:
         return href + "/" + suffix, brain
```
What if there is a suffix and an anchor?

sneridagh replied:
Done

sneridagh commented:
I know, it’s one or the other, never both. Do you have a use case in mind?

Víctor Fernández de Alba
Github/Twitter: sneridagh


On Wed, 31 Jan 2024 at 20:37, David Glick ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In src/plone/restapi/deserializer/utils.py
> <https://github.com/plone/plone.restapi/pull/1746#discussion_r1473371669>:
>
> > @@ -25,11 +25,13 @@ def path2uid(context, link):
>          )
>
>      # handle edge-case when we have non traversable path like /@@download/file
> -    if "/@@" in path:
> -        path, suffix = path.split("/@@", 1)
> -        suffix = "/@@" + suffix
> -    else:
> -        suffix = ""
> +    SUFFIXES = ["/@@", "#"]
> +    suffix = ""
> +    for suffix_separator in SUFFIXES:
> +        if suffix_separator in path:
> +            path, suffix = path.split(suffix_separator, 1)
>
> @sneridagh <https://github.com/sneridagh> You are overwriting suffix
> here. So if both separators are present, only one of the suffixes will be
> kept.
> ------------------------------
>
> In src/plone/restapi/serializer/utils.py
> <https://github.com/plone/plone.restapi/pull/1746#discussion_r1473372777>:
>
> > @@ -23,13 +23,15 @@ def resolve_uid(path):
>      if match is None:
>          return path, None
>
> -    uid, suffix = match.groups()
> +    uid, suffix, anchor = match.groups()
>      brain = uuidToCatalogBrain(uid)
>      if brain is None:
>          return path, None
>      href = brain.getURL()
>      if suffix:
>          return href + "/" + suffix, brain
>
> What if there is a suffix and an anchor?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/plone/plone.restapi/pull/1746#pullrequestreview-1854675752>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AADW4D3TZMLO7PPU4CO3EILYRKMOPAVCNFSM6AAAAABCQ2L5ECVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNJUGY3TKNZVGI>
> .
> You are receiving this because you were mentioned.Message ID: <plone/plone
> .***@***.***>
>

sneridagh commented:
@davisagli I think not, because a suffix is something that matches a `/@@` BrowserView. A BrowserView with an anchor, in RESTAPI world? Could be in ClassicLand, but not in RESTAPI. We could make it work, but I think that without a use case, it's premature optimization, could be that there's never a use case for that.

davisagli commented:
@sneridagh Although this is plone.restapi, the URLs that an editor can enter are not API URLs. It could be a browser view that renders HTML including anchors, sure. Here we need to be careful and pass through unchanged anything that is not the path to a content item.

davisagli commented on news/1746.bugfix:
```diff
@@ -0,0 +1 @@
+Improve the block transforms to include also anchors when resolving uid back and forth @sneridagh
```
```suggestion
Fixed the logic for converting public URLs to and from internal UID-based URLs. Now if the URL includes a fragment, it is preserved. @sneridagh
```

davisagli reviewed (changes requested):
Uh, the tests are failing...

davisagli reviewed (approved):
Okay, let's see what jenkins says...

davisagli commented:
@jenkins-plone-org please run jobs
